### PR TITLE
Add example with rabbitmq hub impl

### PR DESCRIPTION
**Pull request by aashikam**
## Purpose
Fixes: https://github.com/ballerina-platform/ballerina-standard-library/issues/1713

## Checklist
- [x] Linked to an issue
- [ ] Updated the changelog
- [ ] Added tests

### CODE REVIEW

**Comment by shafreenAnfar** [comment]
We decided to remove the consolidater and keep the implementation simple. We need to update the image accordingly as well.

File: examples/rabbitmq-hub/subscriber/subscriber_service.bal [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/websub;
+import ballerina/log;
+
+listener websub:Listener securedSubscriber = new(9100, host = "localhost");
```
Comment (ayeshLK) [inline examples/rabbitmq-hub/subscriber/subscriber_service.bal:20]:
```suggestion
listener websub:Listener securedSubscriber = new(9100);
```

File: examples/rabbitmq-hub/subscriber/subscriber_service.bal [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+@websub:SubscriberServiceConfig {
+    target: ["http://localhost:9000/hub", "test"], leaseSeconds: 36000
```
Comment (ayeshLK) [inline examples/rabbitmq-hub/subscriber/subscriber_service.bal:23]:
```suggestion
    target: ["http://localhost:9000/hub", "test"], 
    leaseSeconds: 36000
```

File: examples/rabbitmq-hub/publisher/topic_registration_client.bal [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/websubhub;
+import ballerina/io;
+
+public function main() returns error? {
+    websubhub:PublisherClient websubHubClientEP = check new("http://localhost:9000/hub");
+    var registrationResponse = websubHubClientEP->registerTopic("test");
```
Comment (ayeshLK) [inline examples/rabbitmq-hub/publisher/topic_registration_client.bal:22]:
```suggestion
    websubhub:TopicRegistrationSuccess registrationResponse = check websubHubClientEP->registerTopic("test");
```

File: examples/rabbitmq-hub/hub/start_hub.bal [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+        while true {
+            websubhub:VerifiedSubscription[]? persistedSubscribers = check getPersistedSubscribers();
+            if persistedSubscribers is websubhub:VerifiedSubscription[] {
+                refreshSubscribersCache(persistedSubscribers);
+                check startMissingSubscribers(persistedSubscribers);
+            }
+        }
+    } on fail var e {
+        return e;
+    }
```
Comment (ayeshLK) [inline examples/rabbitmq-hub/hub/start_hub.bal:97]:
```suggestion
    while true {
        websubhub:VerifiedSubscription[]? persistedSubscribers = check getPersistedSubscribers();
        if persistedSubscribers is websubhub:VerifiedSubscription[] {
            refreshSubscribersCache(persistedSubscribers);
            check startMissingSubscribers(persistedSubscribers);
        }
    }
```

File: examples/rabbitmq-hub/hub/start_hub.bal [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+    do {
+        while true {
+            websubhub:TopicRegistration[]? persistedTopics = check getPersistedTopics();
+            if persistedTopics is websubhub:TopicRegistration[] {
+                refreshTopicCache(persistedTopics);
+            }
+        }
+    } on fail var e {
+        return e;
+    }
```
Comment (ayeshLK) [inline examples/rabbitmq-hub/hub/start_hub.bal:50]:
```suggestion
    while true {
        websubhub:TopicRegistration[]? persistedTopics = check getPersistedTopics();
        if persistedTopics is websubhub:TopicRegistration[] {
            refreshTopicCache(persistedTopics);
        }
    }
```

File: examples/rabbitmq-hub/hub/hub_service.bal [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+            }
+        }
+    }
+
+    # Processes a verified subscription request.
+    # 
+    # + message - Details of the subscription
+    # + return - `error` if there is any unexpected error or else `()`
+    isolated remote function onSubscriptionIntentVerified(websubhub:VerifiedSubscription message) returns error? {
+        string groupName = util:generateGroupName(message.hubTopic, message.hubCallback);
```
Comment (ayeshLK) [inline examples/rabbitmq-hub/hub/hub_service.bal:106]:
```suggestion
```

File: examples/rabbitmq-hub/hub/hub_service.bal [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+                                + message.hubTopic + "] and Callback [" + message.hubCallback + "]");
+            }
+        }       
+    }
+
+    # Processes a verified unsubscription request.
+    # 
+    # + message - Details of the unsubscription
+    isolated remote function onUnsubscriptionIntentVerified(websubhub:VerifiedUnsubscription message) {
+        string groupName = util:generateGroupName(message.hubTopic, message.hubCallback);
```
Comment (ayeshLK) [inline examples/rabbitmq-hub/hub/hub_service.bal:145]:
```suggestion
```

**Comment by ayeshLK** [comment]
Do we need to add a `.gitignore` for the example ?

**Comment by aashikam** [comment]
> Do we need to add a `.gitignore` for the example ?

I added because I remembered like I added it in other examples. Seems unnecessary. Removed the file.

File: examples/rabbitmq-hub/hub/hub_service.bal [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+import rabbitmqHub.persistence as persist;
+import rabbitmqHub.util;
+
+websubhub:Service hubService =
+service object {
+
+    # Registers a `topic` in the hub.
+    # 
+    # + message - Details related to the topic-registration
+    # + headers - `http:Headers` of the original `http:Request`
```
Comment (ayeshLK) [inline examples/rabbitmq-hub/hub/hub_service.bal:29]:
```suggestion
```

File: examples/rabbitmq-hub/hub/hub_service.bal [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+            if persistingResult is error {
+                log:printError("Error occurred while persisting the topic-registration ", err = persistingResult.message());
+            }
+        }
+    }
+
+    # De-registers a `topic` in the hub.
+    # 
+    # + message - Details related to the topic-deregistration
+    # + headers - `http:Headers` of the original `http:Request`
```
Comment (ayeshLK) [inline examples/rabbitmq-hub/hub/hub_service.bal:54]:
```suggestion
```

File: examples/rabbitmq-hub/hub/hub_service.bal [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+        } 
+    }
+
+    # Publishes content to the hub.
+    # 
+    # + message - Details of the published content
+    # + headers - `http:Headers` of the original `http:Request`
+    # + return - `websubhub:Acknowledgement` if publish content is successful, `websubhub:UpdateMessageError`
+    #            if publish content failed or `error` if there is any unexpected error
+    isolated remote function onUpdateMessage(websubhub:UpdateMessage message, http:Headers headers)
```
Comment (ayeshLK) [inline examples/rabbitmq-hub/hub/hub_service.bal:158]:
```suggestion
    # Publishes content to the hub.
    # 
    # + message - Details of the published content
    # + return - `websubhub:Acknowledgement` if publish content is successful, `websubhub:UpdateMessageError`
    #            if publish content failed or `error` if there is any unexpected error
    isolated remote function onUpdateMessage(websubhub:UpdateMessage message)
```

File: examples/rabbitmq-hub/hub/hub_service.bal [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/websubhub;
+import ballerina/log;
+import ballerina/http;
```
Comment (ayeshLK) [inline examples/rabbitmq-hub/hub/hub_service.bal:19]:
```suggestion
```

File: examples/rabbitmq-hub/README.md [hunk 12]
```diff
@@ -0,0 +1,76 @@
+# A Websub Hub Based on RabbitMQ
+
+# Overview
+
+This is a websub hub implementation is based on message brokers. In this case we have used RabbitMQ message broker to implement the websubhub APIs. Message brokers are a natural fit for the implementation of websub hubs as they serve a similar purpose. However, regardless of the APIs used to publish messages to the hub, websubhub APIs can be used to implement websub hubs which are fully compliant with the websub specification.
```
Comment (ayeshLK) [inline examples/rabbitmq-hub/README.md:5]:
```suggestion
The purpose of this example is to demonstrate the usage of ballerina RabbbitMQ package in a production ready application. 

This is a websub hub implementation based on message brokers. In this case hub is backed by RabbitMQ message broker. Message brokers are a natural fit for the implementation of websub hubs as they serve a similar purpose.
```

File: examples/rabbitmq-hub/hub/Config.toml [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+
+# RabbitMQ queue which will get notified for websub subscription/unsubscription
+# All the hubs must be pointed to the same RabbitMQ queue to notify websub subscription/unsubscription
+WEBSUB_SUBSCRIBERS_QUEUE = "registered-websub-subscribers"
+
+# The interval in which RabbitMQ consumers wait for new messages
+POLLING_INTERVAL = 10.0
+
+# The period in which RabbitMQ close method waits to complete
+GRACEFUL_CLOSE_PERIOD = 5.0
```
Comment (ayeshLK) [inline examples/rabbitmq-hub/hub/Config.toml:18]: Do we need these configurations ?

**Comment by ayeshLK** [comment]
Since `Dependencies.toml` is an auto-generated artifact shall we ignore the `Dependencies.toml` of examples by adding following line to repository `.gitignore` file.
```sh
examples/**/Dependencies.toml
```

**Review by ayeshLK — APPROVED** [review]

**Pull request merged**